powerfulseal/clouddrivers/aws_driver.py

```diff
@@ -0,0 +1,90 @@
+import logging
+import boto3
+from . import AbstractDriver
+from ..node import Node, NodeState
+
+
+def create_connection_from_config():
+    """ Creates a new aws api connection """
+    conn = boto3.resource('ec2')
+    return conn
+
+def get_all_ips(instance):
+    """
+        Digs out all the IPs from the instance.private_ip_address field
+        of an AWS EC2 instances
+    """
+    output = []
+    output.append(instance.private_ip_address)
+    output.append(instance.public_ip_address)
+    return output
+
+#https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceState.html
+MAPPING_STATES_STATUS = {
+    "RUNNING": NodeState.UP,
+    "STOPPED": NodeState.DOWN,
+    "TERMINATED": NodeState.DOWN,
+}
+
+def server_status_to_state(status):
+    return MAPPING_STATES_STATUS.get(status['Name'].upper(), NodeState.UNKNOWN)
+
+def create_node_from_server(server, ip):
+    """ Translate AWS EC2 Instance representation into a Node object.
+    """
+    return Node(
+        id=server.id,
+        ip=ip,
+        az=server.placement['AvailabilityZone'],
+        name="",
+        state=server_status_to_state(server.state),
+)
+
+class AWSDriver(AbstractDriver):
+    """
+        Concrete implementation of the AWS cloud driver.
+    """
+
+    def __init__(self, cloud=None, conn=None, logger=None):
+        self.logger = logger or logging.getLogger(__name__)
+        self.conn = create_connection_from_config()
+        self.instances = []
+
+    def sync(self):
+        """ Downloads a fresh set of nodes form the API.
+        """
+        self.logger.info("Synchronizing remote nodes")
+        """ Downloads a fresh set of nodes form the API.
+        """
+        self.logger.info("Synchronizing remote nodes")
+        self.remote_servers = self.conn.instances.all()
+        self.amount_of_servers = sum(1 for _ in self.remote_servers)
```

Comment:
This doesn't seem to be used anywhere apart from the line just below. How would you feel about swapping this for something like this:

```py
self.remote_servers = self.conn.instances.all()
self.logger.info("Fetched %s remote servers" % len(self.remote_servers))
```

Reply:
self.remote_servers = self.conn.instances.all() this is Collection, which doesn't have method `len()` :(

Reply:
I see...

[The docs](http://boto3.readthedocs.io/en/latest/guide/collections.html) seem to be pointing that the collection does the behind-the-scenes smarts like pagination and reading on demand. Since after sync we expect all of them to be loaded, maybe it's easier to transform the collection into a list `list(self.conn.instances.all())` and be done with it ? What do you reckon ?

Also, happy Monday ! 😄